src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs

```diff
@@ -155,18 +157,49 @@ private async Task ReadBody()
     private async Task WriteBody(bool flush = false)
     {
         Exception? error = null;
+
         try
         {
             while (true)
             {
                 var result = await _bodyOutput.Reader.ReadAsync();
 
                 var buffer = result.Buffer;
+
                 try
                 {
                     if (!buffer.IsEmpty)
                     {
-                        await AsyncIO!.WriteAsync(buffer);
+                        if (buffer.IsSingleSegment)
+                        {
+                            await AsyncIO!.WriteAsync(buffer);
+                        }
+                        else
+                        {
+                            ushort chunksCount = 0;
+                            var start = 0;
+                            var length = 0;
+
+                            foreach (var chunks in buffer)
+                            {
+                                chunksCount++;
+                                length += chunks.Length;
+
+                                if (chunksCount == RESPONSE_BODY_MAX_CHUNKS)
+                                {
+                                    await AsyncIO!.WriteAsync(buffer.Slice(start, length));
+
+                                    chunksCount = 0;
+                                    start = length;
+                                    length = 0;
+                                }
+                            }
+
+                            if (chunksCount > 0)
+                            {
+                                await AsyncIO!.WriteAsync(buffer.Slice(start, length));
+                            }
+                        }
```

Comment:
Can you move this logic into WriteAsync instead.

Reply:
I see what I can do.